src/baldr/tilehierarchy.cc

```diff
@@ -1,44 +1,100 @@
+                midgard::Tiles<midgard::PointLL>{wholemapbb.minpt(), tilesizesvector[0],
+                                                 static_cast<int32_t>(std::round(wholemapbb.Width() /
+                                                                                 tilesizesvector[0])),
+                                                 static_cast<int32_t>(std::round(wholemapbb.Height() /
+                                                                                 tilesizesvector[0])),
                                                  static_cast<unsigned short>(kBinsDim)}},
 
       TileLevel{1, stringToRoadClass("Tertiary"), "arterial",
-                midgard::Tiles<midgard::PointLL>{{{-180, -90}, {180, 90}},
-                                                 1,
+                midgard::Tiles<midgard::PointLL>{wholemapbb.minpt(), tilesizesvector[1],
+                                                 static_cast<int32_t>(std::round(wholemapbb.Width() /
+                                                                                 tilesizesvector[1])),
+                                                 static_cast<int32_t>(std::round(wholemapbb.Height() /
+                                                                                 tilesizesvector[1])),
                                                  static_cast<unsigned short>(kBinsDim)}},
 
       TileLevel{2, stringToRoadClass("ServiceOther"), "local",
-                midgard::Tiles<midgard::PointLL>{{{-180, -90}, {180, 90}},
-                                                 .25,
+                midgard::Tiles<midgard::PointLL>{wholemapbb.minpt(), tilesizesvector[2],
+                                                 static_cast<int32_t>(std::round(wholemapbb.Width() /
+                                                                                 tilesizesvector[2])),
+                                                 static_cast<int32_t>(std::round(wholemapbb.Height() /
+                                                                                 tilesizesvector[2])),
                                                  static_cast<unsigned short>(kBinsDim)}},
+
   };
 
+  transit_levels_ = {
+      TileLevel{3, stringToRoadClass("ServiceOther"), "transit",
+                midgard::Tiles<midgard::PointLL>{wholemapbb.minpt(), tilesizesvector[3],
+                                                 static_cast<int32_t>(std::round(wholemapbb.Width() /
+                                                                                 tilesizesvector[3])),
+                                                 static_cast<int32_t>(std::round(wholemapbb.Height() /
+                                                                                 tilesizesvector[3])),
+                                                 static_cast<unsigned short>(kBinsDim)}},
+  };
+}
+
+const std::vector<TileLevel>& TileHierarchy::levels() {
+  if (levels_.size() == 0) {
+    getLevelsfromConfig();
```

Comment:
getting these values from config doesnt seem like the right idea to me. as @dnesbitt61 it invites the possibility of having a config that doesnt match the actual tiles.

if we want to make this configurable, we have to:

1. validate the config doesnt break the rules (evenly divisble adjacent levels)
2. the config is only used to build tiles (ie it is a mjolnir config)
3. the service and library discover the tilesize from the tiles themselves via graphtileheader

Reply:
that last one is a deal breaker in that it would require a major version change. we would have to add something to the graphtileheader and add code to graphreader et al to read that. we can do that as a new feature in valhalla v3 but if we make tiles that arent using the default sizes and old valhalla v3 code wants to read those, it will break because it wont know anything about the configurable sizes and having to get that info from graphtileheader. sadly to actually do this we need to do it in a new major version

Reply:
Yes, old valhalla v3 code would not be able to use the new tiles. Can't we handle this feature as a new, experimental one and warn the users about this ? 

@kevinkreiser , could you advice how to progress with this ?
-  Should this PR be cancelled now, and later the design/implementation of the final graphtile-embedded tilesizes can be discussed, when preparing to switch to v4.
 - Or, would it worth to include the configurable tilesize feature in some form (based on this PR) in the current, v3 series as a temporary solution, so, at least we could start experimenting it ?

Reply:
Hm, but this modification doesn't solve the initial problem of breaking a consuming application if the tiles were built non-standard, right? Then what's the point if there can't be a graph reader being able to read them?

> Thanks, we always wanted to have configurable tile sizes.

I was very confused here haha I thought about hierarchy limits, not tile dimensions.

Reply:
That's true, it does not solve the compatibility issue, but, at least it enables this experimental feature in v3. I don't know about the v4 release plans, but if it's in the far future, would it not be better to give some possibility to experiment with this feature ?

Reply:
Well, there would be 2 use cases:

- I believe all future releases from master would be able to convert OSM maps with a custom layout if this is merged.  Am I wrong ?
- Also, we could produce graph tiles from other sources with custom layout.

The feature is on our forked branch, we can use it, so I won't feel bad if you refuse to accept this contribution.

Reply:
Jep then you could _build_ tiles that way, but you won’t be able to _read_ the same tiles right? The graph reader will still expect the default resolution. Or am I missing smth?

Reply:
Nope, I could produce graph tiles using a custom tile grid, and use the valhalla-app viewer/tester to calculate routes on it when I specified the same custom layout in the valhalla.json.
So it seems that the graph reader and the whole valhalla_service used this custom layout correctly.

Reply:
Huh interesting.. then I’d be clearly missing smth😄